### PR TITLE
Make HubConnection write messages directly to the PipeWriter

### PR DESCRIPTION
**Pull request by JamesNK**
Follow up on https://github.com/aspnet/SignalR/issues/1633, possible now that the connection refactor is done.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs [hunk 1]
```diff
@@ -296,7 +296,7 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
                     request.Version = new Version(1, 1);
                     SendUtils.PrepareHttpRequest(request, _httpOptions);
 
-                    using (var response = await httpClient.SendAsync(request))
+                    using (var response = await httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead))
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: @davidfowl, what is a comment to put with this setting to explain what it optimizes?

Comment (JamesNK) [reply src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs]: nvm, looked up what it did

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionSendBenchmark.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        public void GlobalCleanup()
+        {
+            _tcs.SetResult(new ReadResult(default, false, true));
+            _hubConnection.StopAsync().GetAwaiter().GetResult();
+        }
+
+        [Benchmark]
+        public Task SendAsync()
+        {
+            return _hubConnection.SendAsync("Dummy", Array.Empty<object>());
```
Comment (davidfowl) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/HubConnectionSendBenchmark.cs]: Can you do a params with a couple of different arg types, empty, one and two. Not sure about sizes, we can look at that later.

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 3]
```diff
@@ -343,11 +343,14 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
...
 
-            var payload = _protocol.WriteToArray(hubMessage);
+            PipeWriterStream pipeWriterStream = new PipeWriterStream(_connectionState.Connection.Transport.Output);
+            _protocol.WriteMessage(hubMessage, pipeWriterStream);
 
             Log.SendingMessage(_logger, hubMessage);
+
             // REVIEW: If a token is passed in and is cancelled during FlushAsync it seems to break .Complete()...
-            await WriteAsync(payload, CancellationToken.None);
+            await _connectionState.Connection.Transport.Output.FlushAsync(CancellationToken.None);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: Just don't pass the connection token.

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 4]
```diff
@@ -343,11 +343,14 @@ public IDisposable On(string methodName, Type[] parameterTypes, Func<object[], o
         {
             AssertConnectionValid();
 
-            var payload = _protocol.WriteToArray(hubMessage);
+            PipeWriterStream pipeWriterStream = new PipeWriterStream(_connectionState.Connection.Transport.Output);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: We should create this once and stash it on the connection state.

**Comment by JamesNK** [comment]
Before

```
    Method | ArgumentCount | Protocol |      Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
---------- |-------------- |--------- |----------:|----------:|----------:|----------:|-------:|----------:|
 SendAsync |             0 |     json |  3.662 us | 0.0727 us | 0.1964 us | 273,075.8 | 0.6828 |    6216 B |
 SendAsync |             0 |  msgpack |  2.714 us | 0.1091 us | 0.3183 us | 368,441.7 | 0.0725 |     712 B |
 SendAsync |             1 |     json |  4.296 us | 0.1105 us | 0.3099 us | 232,780.1 | 0.7019 |    6352 B |
 SendAsync |             1 |  msgpack |  2.865 us | 0.0543 us | 0.0646 us | 348,990.7 | 0.0916 |     904 B |
 SendAsync |            10 |     json |  6.289 us | 0.1349 us | 0.2693 us | 159,005.1 | 0.8240 |    7560 B |
 SendAsync |            10 |  msgpack |  6.071 us | 0.0924 us | 0.0819 us | 164,725.7 | 0.2670 |    2600 B |
 SendAsync |           100 |     json | 29.737 us | 0.1737 us | 0.1356 us |  33,627.6 | 1.8311 |   19713 B |
 SendAsync |           100 |  msgpack | 37.406 us | 0.4400 us | 0.4116 us |  26,734.0 | 1.7090 |   19617 B |
```

After

```
    Method | ArgumentCount | Protocol |      Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
---------- |-------------- |--------- |----------:|----------:|----------:|----------:|-------:|----------:|
 SendAsync |             0 |     json |  3.374 us | 0.1137 us | 0.3131 us | 296,402.9 | 0.6790 |    6128 B |
 SendAsync |             0 |  msgpack |  2.401 us | 0.0475 us | 0.1292 us | 416,425.2 | 0.0687 |     624 B |
 SendAsync |             1 |     json |  3.674 us | 0.0728 us | 0.1470 us | 272,176.8 | 0.6943 |    6248 B |
 SendAsync |             1 |  msgpack |  2.880 us | 0.0576 us | 0.1214 us | 347,254.0 | 0.0877 |     800 B |
 SendAsync |            10 |     json |  5.875 us | 0.1121 us | 0.0994 us | 170,224.2 | 0.8011 |    7328 B |
 SendAsync |            10 |  msgpack |  5.707 us | 0.1123 us | 0.1201 us | 175,224.3 | 0.2441 |    2384 B |
 SendAsync |           100 |     json | 27.400 us | 0.3050 us | 0.2381 us |  36,496.8 | 1.7395 |   18129 B |
 SendAsync |           100 |  msgpack | 37.002 us | 0.7245 us | 0.8625 us |  27,025.6 | 1.7090 |   18225 B |
```

**Pull request merged**